retrievalmarket/network/libp2p_impl_test.go

```diff
@@ -0,0 +1,516 @@
+			drChan <- readDP
+		},
+	}
+	require.NoError(t, toNetwork.SetDelegate(tr2))
+	assertDealResponseReceived(ctx, t, fromNetwork, toPeer, drChan)
+}
+
+func TestDealStreamSendReceiveDealPayment(t *testing.T) {
+	// send payment, read in handler
+	ctx := context.Background()
+	td := shared_testutil.NewLibp2pTestData(ctx, t)
+	fromNetwork := network.NewFromLibp2pHost(td.Host1)
+	toNetwork := network.NewFromLibp2pHost(td.Host2)
+	toPeer := td.Host2.ID()
+
+	tr := &testReceiver{t: t}
+	require.NoError(t, fromNetwork.SetDelegate(tr))
+
+	dpyChan := make(chan retrievalmarket.DealPayment)
+	tr2 := &testReceiver{
+		t: t,
+		dealStreamHandler: func(s network.RetrievalDealStream) {
+			readDpy, err := s.ReadDealPayment()
+			require.NoError(t, err)
+			dpyChan <- readDpy
+		},
+	}
+	require.NoError(t, toNetwork.SetDelegate(tr2))
+	assertDealPaymentReceived(ctx, t, fromNetwork, toPeer, dpyChan)
+}
+
+func TestDealStreamSendReceiveMultipleSuccessful(t *testing.T) {
+	// send proposal, read in handler, send response back,
+	// read response,
+	// send payment, read farther in handler
+
+	bgCtx := context.Background()
+	td := shared_testutil.NewLibp2pTestData(bgCtx, t)
+	fromNetwork := network.NewFromLibp2pHost(td.Host1)
+	toNetwork := network.NewFromLibp2pHost(td.Host2)
+	toPeer := td.Host2.ID()
+
+	// set up stream handler, channels, and response
+	// dpChan := make(chan retrievalmarket.DealProposal)
+	dpyChan := make(chan retrievalmarket.DealPayment)
+	dr := shared_testutil.MakeTestDealResponse()
+
+	tr2 := &testReceiver{t: t, dealStreamHandler: func(s network.RetrievalDealStream) {
+		_, err := s.ReadDealProposal()
+		require.NoError(t, err)
+
+		require.NoError(t, s.WriteDealResponse(dr))
+
+		readDp, err := s.ReadDealPayment()
+		require.NoError(t, err)
+		dpyChan <- readDp
+	}}
+	require.NoError(t, toNetwork.SetDelegate(tr2))
+
+	// start sending deal proposal
+	ds1, err := fromNetwork.NewDealStream(toPeer)
+	require.NoError(t, err)
+
+	dp := shared_testutil.MakeTestDealProposal()
+
+	var receivedPayment retrievalmarket.DealPayment
+
+	ctx, cancel := context.WithTimeout(bgCtx, 10*time.Second)
+	defer cancel()
+
+	// write proposal
+	require.NoError(t, ds1.WriteDealProposal(dp))
+
+	// read response and verify it's the one we told toNetwork to send
+	responseReceived, err := ds1.ReadDealResponse()
+	require.NoError(t, err)
+	assert.Equal(t, dr.ID, responseReceived.ID)
+	assert.Equal(t, dr.Message, responseReceived.Message)
+	assert.Equal(t, dr.Status, responseReceived.Status)
+
+	// send payment
+	dpy := retrievalmarket.DealPayment{
+		ID:             dp.ID,
+		PaymentChannel: address.TestAddress,
+		PaymentVoucher: shared_testutil.MakeTestSignedVoucher(),
+	}
+	require.NoError(t, ds1.WriteDealPayment(dpy))
+
+	select {
+	case <-ctx.Done():
+		t.Errorf("failed to receive messages")
+	case receivedPayment = <-dpyChan:
+	}
+
+	assert.Equal(t, dpy, receivedPayment)
+}
+
+func TestQueryStreamSendReceiveMultipleOutOfOrderFails(t *testing.T) {
```

Comment:
This is my own fault, but, honestly, this test doesn't make sense in retrospect.

I think the reason you're deadlocking without a go routine is the second WriteDealProposal never finishes in side the go routine. But cause there's no channel write at the end of the go routine, you just never see that.

It would be interesting to put a channel write at the end of the go routine and then try to read from it at the very end. I think you might find it never gets written to.

Assuming that's the case, I think this test gives a false impression of testing something it doesn't. Namely that all the writes after the first do not in fact pass. 

I'm not sure if there's a good way to resolve this, and I wonder if it's just unrealstic to trying to have this test -- in practice,
the handlers we write are likely to return as soon as they encounter an error, and there will not be multiple error reads in a row. 

So my honest, though non-blocking, recommendation is to delete this test and simplify the out-of-order query read test above to just test a single write/read, so that we don't have to deal with this. Sorry for having unneccesary work.